gossipsub.go

```diff
@@ -239,6 +264,90 @@ func (gs *GossipSubRouter) handlePrune(p peer.ID, ctl *pb.ControlMessage) {
 			gs.tracer.Prune(p, topic)
 			delete(peers, p)
 			gs.untagPeer(p, topic)
+			gs.addBackoff(p, topic)
+			px := prune.GetPeers()
+			if len(px) > 0 {
+				gs.pxConnect(px)
+			}
+		}
+	}
+}
+
+func (gs *GossipSubRouter) addBackoff(p peer.ID, topic string) {
+	backoff, ok := gs.backoff[topic]
+	if !ok {
+		backoff = make(map[peer.ID]time.Time)
+		gs.backoff[topic] = backoff
+	}
+	backoff[p] = time.Now().Add(GossipSubPruneBackoff)
+}
+
+func (gs *GossipSubRouter) pxConnect(peers []*pb.PeerInfo) {
+	toconnect := make([]connectInfo, 0, len(peers))
+	for _, pi := range peers {
+		p := peer.ID(pi.PeerID)
+
+		_, connected := gs.peers[p]
+		if connected {
+			continue
+		}
+
+		var srr *routing.SignedRoutingState
+		var err error
+		if pi.SignedAddrs != nil {
+			// the peer sent us a signed record; ensure that it is valid
+			srr, err = routing.UnmarshalSignedRoutingState(pi.SignedAddrs)
+			if err != nil {
+				log.Warningf("error unmarshalling routing record obtained through px: %s", err)
+				continue
+			}
+			if srr.PeerID != p {
+				log.Warningf("bogus routing record obtained through px: peer ID %s doesn't match expected peer %s", srr.PeerID, p)
+				continue
+			}
+		}
+
+		toconnect = append(toconnect, connectInfo{p, srr})
+	}
+
+	if len(toconnect) == 0 {
+		return
+	}
+
+	// initiate connections, without blocking the event loop
+	go func() {
+		for _, ci := range toconnect {
+			select {
+			case gs.connect <- ci:
```

Comment:
I'm concerned about how many goroutines may end up being stuck here if the channel is backed up due to many ongoing slow DHT requests (especially if we end up looking for peers that don't exist/are not dialable).

Two approaches that may help here are:
1) Use a backoff connector object like we have in the discovery module (perhaps even move that object into pubsub so that gossipsub can share it). This will help with minimizing repeated dials to bad peers.
2) Use an infinite buffer (e.g. a map protected by a mutex). I'd rather an infinite buffer than infinite goroutines any day of the week

Reply:
Note that the DHT requests should all eventually disappear as gossipsub v1.1 gets deployed, because we'll have signed address record from all peers.

Reply:
Honestly I don't think this is an issue, and it will greatly complicate code to use a mutex/map combo.

Reply:
Also note that the channel should be drained pretty fast, and that these connection attempts are not all that frequent.

Reply:
I agree they'll be drained fast if we have signed peer exchange with good NAT punchthrough. It could get clogged up though if we're doing DHT queries or bad dials.

Frequency is hopefully low, although I wonder if applications might get in trouble during startup. If we use Discover to find more than DHi peers on initialization and connect to them then if the application starts by connecting to 20 topics we may end up with many stalled goroutines.

Since it's not blocking the event loop, I'm ok with a wait + see approach. We may need to keep an eye on this and
/or write out a testground test to exercise this

Reply:
agreed.

Reply:
Also note, many stalled goroutines is not that big of a deal -- the real issue is unbounded goroutine buildup, but I don't think this will happen.